src/particles/particle_creation.hpp

```diff
@@ -135,6 +139,241 @@ template <ParticleType particleType> struct ParticleCreationTraits {
+				const amrex::Real mass_low_mass_star = particle_mass * (1.0 - fstar_high);
+				double total_momx = 0.0;
+				double total_momy = 0.0;
+				double total_momz = 0.0;
+
+				for (int p_idx = 0; p_idx < num_particles; ++p_idx) {
+					auto &p = particles[p_idx]; // NOLINT
+
+					// Set particle ID and CPU
+					p.id() = pid_start + base_offset + p_idx;
+					p.cpu() = cpu_id;
+					p.rdata(birth_time_index) = current_time;
+
+					// Set particle evolution stage
+					p.idata(evolution_stage_index) = p_idx == 0 ? static_cast<int>(StellarEvolutionStage::LowMassStar)
+										    : static_cast<int>(StellarEvolutionStage::SNProgenitor);
+
+					// Low Mass particle position at cell center
+					p.pos(0) = plo[0] + (i + 0.5) * dx[0];
+					p.pos(1) = plo[1] + (j + 0.5) * dx[1];
+					p.pos(2) = plo[2] + (k + 0.5) * dx[2];
+
+					// Low Mass particle mass and velocity
+					p.rdata(mass_idx) = mass_low_mass_star;
+					p.rdata(mass_idx + 1) = vx;
+					p.rdata(mass_idx + 2) = vy;
+					p.rdata(mass_idx + 3) = vz;
+
+					p.rdata(birth_time_index + 1) = LONG_MAX;
+					if (p_idx > 0) {
+						double sigma_sq_x = NAN;
+						double sigma_sq_y = NAN;
+						double sigma_sq_z = NAN;
+						double numx = 0.0;
+						double numy = 0.0;
+						double numz = 0.0;
+						double denominator = 0.0;
+						double vx_adj = NAN;
+						double vy_adj = NAN;
+						double vz_adj = NAN;
+						double rho_adj = NAN;
+						// Get the average velocity from the adjoining cells
+						for (int ii = i - 1; ii <= i + 1; ++ii) {
+							for (int jj = j - 1; jj <= j + 1; ++jj) {
+								for (int kk = k - 1; kk <= k + 1; ++kk) {
+
+									vx_adj = (state_arr(ii, jj, kk, HydroSystem<problem_t>::x1Momentum_index)) /
+										 state_arr(ii, jj, kk, HydroSystem<problem_t>::density_index);
+									vy_adj = (state_arr(ii, jj, kk, HydroSystem<problem_t>::x2Momentum_index)) /
+										 state_arr(ii, jj, kk, HydroSystem<problem_t>::density_index);
+									vz_adj = (state_arr(ii, jj, kk, HydroSystem<problem_t>::x3Momentum_index)) /
+										 state_arr(ii, jj, kk, HydroSystem<problem_t>::density_index);
+									rho_adj = state_arr(ii, jj, kk, HydroSystem<problem_t>::density_index);
+									numx += rho_adj * (vx_adj - (vx)) * (vx_adj - (vx));
+									numy += rho_adj * (vy_adj - (vy)) * (vy_adj - (vy));
+									numz += rho_adj * (vz_adj - (vz)) * (vz_adj - (vz));
+
+									denominator += rho_adj;
+								}
+							}
+						}
+						sigma_sq_x = numx / denominator;
+						sigma_sq_y = numy / denominator;
+						sigma_sq_z = numz / denominator;
+
+						p.rdata(mass_idx + 1) = (std::abs(vx) / vx) * amrex::RandomNormal(std::abs(vx), std::sqrt(sigma_sq_x), engine);
+						p.rdata(mass_idx + 2) = (std::abs(vy) / vy) * amrex::RandomNormal(std::abs(vy), std::sqrt(sigma_sq_y), engine);
+						p.rdata(mass_idx + 3) = (std::abs(vz) / vz) * amrex::RandomNormal(std::abs(vz), std::sqrt(sigma_sq_z), engine);
+
```

Comment:
I don't understand why you need a loop here. You can just draw a single value and guarantee that the mass falls between min and max.